File: ts-sdk/src/metrics/metrics.manager.ts
import { ChainMetrics, MetricsData } from ".";

export class MetricsManager {
  constructor(
    private readonly metrics: Map<number, MetricsData> = new Map() // TODO use lru-cache
  ) {}

  public flushMetrics(): ChainMetrics[] {
    const chainMetrics: ChainMetrics[] = [];
    for (const [chainId, chainIdMetrics] of this.metrics.entries()) {
      const dataPoints: { [metricName: string]: number[] } = {};
      for (const [metricName, metricData] of chainIdMetrics.entries()) {
        dataPoints[metricName] = metricData;
      }
      chainMetrics.push({ chainId, dataPoints });
    }
    this.metrics.clear();
    return chainMetrics;
  }

  public reportMetric(chainId: number, metric: string, value: number) {
    if (!this.metrics.has(chainId)) {
      this.metrics.set(chainId, new Map());
    }

    if (!this.metrics.get(chainId)!.has(metric)) {
      this.metrics.get(chainId)!.set(metric, [value]);
    } else {
      this.metrics.get(chainId)!.get(metric)!.push(value);
    }
  }
}

File: py-sdk/src/forta_bot/metrics/metrics_manager.py
from .types import MetricsData, ChainMetrics
from .constants import MetricName

class MetricsManager:
  def __init__(self):
    self.metrics: dict[int, MetricsData] = {}
  
  def flush_metrics(self) -> list[ChainMetrics]:
    chain_metrics: list[ChainMetrics] = []
    for chain_id, chain_id_metrics in self.metrics.items():
      data_points: dict[str, list[int]] = {}
      for metric_name, metric_data in chain_id_metrics.items():
        data_points[metric_name] = metric_data
      chain_metrics.append({'chainId': chain_id, 'dataPoints': data_points})
    self.metrics.clear()
    return chain_metrics
  
  def report_metric(self, chain_id: int, metric: str, value: int):
    if type(metric) == MetricName:
      metric = metric.value
      
    if chain_id not in self.metrics:
      self.metrics[chain_id] = {}
    
    if metric not in self.metrics[chain_id]:
      self.metrics[chain_id][metric] = [value]
    else:
      self.metrics[chain_id][metric].append(value)